core/user/user.go

```diff
@@ -0,0 +1,23 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package user
+
+import (
+	"time"
+)
+
+// User describes a user within Juju.
+type User struct {
+	// CreatedAt is the time that the user was created at.
+	CreatedAt time.Time
+
+	// DisplayName is a user friendly name represent the user as.
+	DisplayName string
+
+	// Name is the username of the user.
+	Name string
+
+	// Creator is the username of the user that created this user.
+	Creator string
```

Comment:
I'm not ecstatic with the fact that you pass in a user based on another user name. This feels very fragile. I'd rather be very specific and pass in the UUID of the user. We don't have to do it for this PR and it could be a followup.

Also, I'd say we could change the type to better reflect what it is, something akin to this. Not something we want to do now, but we should be thinking about it.

```go
type OwnedUser struct {
	User
	// CreatorUUID is the assoicated user that created this user.
	CreatorUUID string
}

type User struct {
	// CreatedAt is the time that the user was created at.
	CreatedAt time.Time

	// DisplayName is a user friendly name represent the user as.
	DisplayName string

	// Name is the username of the user.
	Name string
}
```

Reply:
Switched to `CreatorUUID`, and made a note (in my notes) about the `OwnedUser` type and a more "clear types" approach in general.

Reply:
I am not sure we should have made this change yet. It's a hard problem to model like that now given how the api works.

We are also adding more external authn providers to Juju with JAAS and we really need to be able to use the natural types that we are going to get from those tokens etc. I also don't believe tags has the concept of uuid.

For my two cents I think we should leave UUID in the state layer until it makes perfect sense this is what we need.

Just my </two-cents>